share/vss/vss.go

```diff
@@ -0,0 +1,630 @@
+	}
+
+	return a.addResponse(r)
+}
+
+func (a *aggregator) verifyJustification(j *Justification) error {
+	if _, ok := findPub(a.verifiers, j.Index); !ok {
+		return errors.New("vss: index out of bounds in justification")
+	}
+	r, ok := a.responses[j.Index]
+	if !ok {
+		return errors.New("vss: no complaints received for this justification")
+	} else if r.Status != StatusComplaint {
+		return errors.New("vss: justification received for an approval")
+	}
+
+	if err := a.VerifyDeal(j.Deal, false); err != nil {
+		// if one response is bad, flag the dealer as malicious
+		a.badDealer = true
+		return err
+	}
+	r.Status = statusJustified
+	return nil
+}
+
+func (a *aggregator) addResponse(r *Response) error {
+	if _, ok := findPub(a.verifiers, r.Index); !ok {
+		return errors.New("vss: index out of bounds in Complaint")
+	}
+	if _, ok := a.responses[r.Index]; ok {
+		return errors.New("vss: already existing response from same origin")
+	}
+	a.responses[r.Index] = r
+	return nil
+}
+
+// EnoughApprovals returns true if enough verifiers have sent their approval for
+// the deal they received.
+func (a *aggregator) EnoughApprovals() bool {
+	var app int
+	for _, r := range a.responses {
+		if r.Status == StatusApproval {
+			app++
+		}
+	}
+	return app >= a.t
+}
+
+// DealCertified returns true if there has been less than t complaints, all
+// Justifications were correct and if EnoughApprovals() returns true.
+func (a *aggregator) DealCertified() bool {
+	var comps int
+	for _, r := range a.responses {
+		if r.Status == StatusComplaint {
+			comps++
+		}
+	}
+	return a.EnoughApprovals() && !(comps >= a.t || a.badDealer)
+}
+
+// MinimumT the minimum safe T that should be used with this protocol as it's
+// proven to be secure with such a T. WARNING: Setting a lower T could make
+// the whole protocol insecure. Setting a higher T only makes it harder to
+// reconstruct the secret.
+func MinimumT(n int) int {
+	return (n + 1) / 2
+}
+
+func validT(t int, verifiers []abstract.Point) bool {
+	return t >= 2 && t <= len(verifiers) && int(uint32(t)) == t
+}
+
+func deriveH(suite abstract.Suite, verifiers []abstract.Point) abstract.Point {
+	var b bytes.Buffer
+	for i := range verifiers {
+		verifiers[i].MarshalTo(&b)
+	}
+	h := suite.Hash()
+	h.Write(b.Bytes())
+	digest := h.Sum(nil)
+	base, _ := suite.Point().Pick(nil, suite.Cipher(digest))
+	return base
+}
+
+func findPub(verifiers []abstract.Point, idx uint32) (abstract.Point, bool) {
+	iidx := int(idx)
+	if iidx >= len(verifiers) {
+		return nil, false
+	}
+	return verifiers[iidx], true
+}
+
+func sessionID(suite abstract.Suite, dealer abstract.Point, verifiers, commitments []abstract.Point, t int) ([]byte, error) {
+	h := suite.Hash()
+	dealer.MarshalTo(h)
+
+	for i := range verifiers {
+		verifiers[i].MarshalTo(h)
+	}
+
+	for i := range commitments {
+		commitments[i].MarshalTo(h)
+	}
+	binary.Write(h, binary.LittleEndian, uint32(t))
+
+	return h.Sum(nil), nil
+}
+
+func msgResponse(r *Response) []byte {
```

Comment:
1. Well I have to sign my messages in a deterministic way so verifiers can produce the same hash, so I would prefer the marshalling to be explicit and visible and language agnostic instead of using some Golang specific construction.
2. Indeed, I hesitated to use a Hash here. Maybe I should, although I have not enough expertise to know whatever it brings more to the table...